False errors due to scope condition being set incorrectly when doing Sema lookup

For example:

```cpp
namespace foo {
  template <typename T>
  void bar() {}
}

void baz() {
#ifdef A
  int var=0;
  for( unsigned i=0; i<10; ++i ) {
    if (i)
      ++var;
  }
#endif
}
```

Produces errors:
```
scope_error_after_template.cpp:10:9: error: i is not defined in context ~A
    if (i)
        ^
scope_error_after_template.cpp:11:9: error: var is not defined in context ~A
      ++var;
        ^
```

These errors are emitted because the scope condition is sometimes set to `True` when it should be `A`. This causes Sema to report that the variable's declaration doesn't cover the entire lookup space.

--- a/lib/Sema/Scope.cpp
+++ b/lib/Sema/Scope.cpp
@@ -93,6 +93,15 @@ void Scope::Init(Scope *parent, unsigned flags) {
   Entity = nullptr;
   ErrorTrap.reset();
   NRVO.setPointerAndInt(nullptr, 0);
+
+  // Set the condition of the new scope to be the same as the parent scope.
+  // This is important. Since scopes are cached we need to ensure that when
+  // cached scopes are retrieved, they aren't left with old conditions
+  if (parent) {
+    setConditional(parent->getConditional());
+  } else {
+    setConditional(new Variability::True);
+  }
 }
 
 bool Scope::containedInPrototypeScope() const {

